File: GeneralMIDI.cc
//
// TextMIDITools Version 1.0.58
//
// Copyright © 2023 Thomas E. Janzen
// License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
// This is free software: you are free to change and redistribute it.
// There is NO WARRANTY, to the extent permitted by law.
//
#if HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */

#include "GeneralMIDI.h"

using namespace std;
using namespace textmidi;
using namespace textmidi::cgm;

//
// These values reflect the MIDI Specification 1.1,
// Chapter 5: General MIDI System Level I.
//
const midi::NumStringMap<cgm::GeneralMIDIGroup> cgm::program_group_map
{
    {"piano",        GeneralMIDIGroup::Piano},
    {"chromaticpercussion", GeneralMIDIGroup::ChromaticPercussion},
    {"organ",        GeneralMIDIGroup::Organ},
    {"guitar",       GeneralMIDIGroup::Guitar},
    {"bass",         GeneralMIDIGroup::Bass},
    {"strings",      GeneralMIDIGroup::Strings},
    {"ensemble",     GeneralMIDIGroup::Ensemble},
    {"brass",        GeneralMIDIGroup::Brass},
    {"reed",         GeneralMIDIGroup::Reed},
    {"pipe",         GeneralMIDIGroup::Pipe},
    {"synthlead",    GeneralMIDIGroup::SynthLead},
    {"synthpad",     GeneralMIDIGroup::SynthPad},
    {"syntheffects", GeneralMIDIGroup::SynthEffects},
    {"ethnic",       GeneralMIDIGroup::Ethnic},
    {"percussive",   GeneralMIDIGroup::Percussive},
    {"soundeffects", GeneralMIDIGroup::SoundEffects},
    {"all",          GeneralMIDIGroup::All},
    {"melodic",      GeneralMIDIGroup::Melodic},
    {"idiophone",    GeneralMIDIGroup::Idiophone}
};

const MIDI_Program cgm::midi_programs[] =
{
    // This is used as a 1-based sequence.
    // MIDI refers to channels and programs in a unit (1)-based way.
    // However elsewhere in this package channel and program are
    // 0-based until they are written to a binary MIDI file.
    { 0, "Dummy",                           {"A0",  "C8"}},
    // Piano
    { 1, "Acoustic Grand Piano",            {"A0",  "C8"}},
    { 2, "Bright Acoustic Piano",           {"A0",  "C8"}},
    { 3, "Electric Grand Piano",            {"E1",  "E7"}},
    { 4, "Honky-tonk Piano",                {"A0",  "C8"}},
    { 5, "Electric Piano 1(Fender Rhodes)", {"A0",  "C8"}},
    { 6, "Electric Piano 2(DX-7 EP)",       {"A0",  "C8"}},
    { 7, "Harpsichord",                     {"F1",  "F6"}},
    { 8, "Clavi",                           {"F1",  "F6"}},
    // Chromatic Percussion
    { 9, "Celesta",                         {"C4",  "C8"}},
    {10, "Glockenspiel",                    {"F5",  "C8"}},
    {11, "Music Box",                       {"F6",  "C8"}},
    {12, "Vibraphone",                      {"F3",  "F6"}},
    {13, "Marimba",                         {"C2",  "C7"}},
    {14, "Xylophone",                       {"C3",  "C7"}},
    {15, "Tubular Bells",                   {"C4",  "G5"}},
    {16, "Dulcimer",                        {"C3",  "C5"}},
    // Organ
    {17, "Drawbar Organ",                   {"A0",  "C8"}},
    {18, "Percussive Organ",                {"A0",  "C8"}},
    {19, "Rock Organ",                      {"A0",  "C8"}},
    {20, "Church Organ",                    {"A0",  "C8"}},
    {21, "Reed Organ",                      {"C2",  "C6"}},
    {22, "Accordion",                       {"E1",  "D6"}},
    {23, "Harmonica",                       {"C3",  "G5"}},
    {24, "Tango Accordion",                 {"E1",  "D6"}},
    // Guitar
    {25, "Acoustic Guitar (nylon)",         {"E3",  "D6"}},
    {26, "Acoustic Guitar (steel)",         {"E3",  "D6"}},
    {27, "Electric Guitar (jazz)",          {"E3",  "D6"}},
    {28, "Electric Guitar (clean)",         {"E3",  "D6"}},
    {29, "Electric Guitar (muted)",         {"E3",  "D6"}},
    {30, "Overdriven Guitar",               {"E3",  "D6"}},
    {31, "Distortion Guitar",               {"E3",  "D6"}},
    {32, "Guitar Harmonics",                {"E4",  "D7"}},
    // Bass
    {33, "Acoustic Bass",                   {"E1",  "E4"}},
    {34, "Electric Bass (finger)",          {"E1",  "E4"}},
    {35, "Electric Bass (pick)",            {"E1",  "E4"}},
    {36, "Fretless Bass",                   {"E1",  "E4"}},
    {37, "Slap Bass 1",                     {"E1",  "E4"}},
    {38, "Slap Bass 2",                     {"E1",  "E4"}},
    {39, "Synth Bass 1",                    {"E1",  "E4"}},
    {40, "Synth Bass 2",                    {"E1",  "E4"}},
    // Strings
    {41, "Violin",                          {"G3",  "C7"}},
    {42, "Viola",                           {"C3",  "D7"}},
    {43, "Cello",                           {"C2",  "A5"}},
    {44, "Contrabass",                      {"E1",  "G4"}},
    {45, "Tremolo Strings",                 {"E1",  "C7"}},
    {46, "Pizzicato Strings",               {"E1",  "C7"}},
    {47, "Orchestral Harp",                 {"B0",  "G#7"}},
    {48, "Timpani",                         {"C2",  "D4"}},
    // Ensemble
    {49, "String Ensemble 1",               {"C0",  "C8"}},
    {50, "String Ensemble 2",               {"C0",  "C8"}},
    {51, "Synth Strings 1",                 {"C0",  "C8"}},
    {52, "Synth Strings 2",                 {"C0",  "C8"}},
    {53, "Choir Aahs",                      {"D2",  "C6"}},
    {54, "Voice Oohs",                      {"D2",  "C6"}},
    {55, "Synth Voice",                     {"D2",  "C6"}},
    {56, "Orchestra Hit",                   {"C0",  "C8"}},
    // Brass
    {57, "Trumpet",                         {"F#3", "C6"}},
    {58, "Trombone",                        {"C2",  "C5"}},
    {59, "Tuba",                            {"Eb2", "A4"}},
    {60, "Muted Trumpet",                   {"F#3", "C6"}},
    {61, "French Horn",                     {"C2",  "C6"}},
    {62, "Brass Section",                   {"C2",  "C6"}},
    {63, "Synth Brass 1",                   {"C2",  "C6"}},
    {64, "Synth Brass 2",                   {"C2",  "C6"}},
    // Reed
    {65, "Soprano Sax",                     {"Ab3", "Eb6"}},
    {66, "Alto Sax", {"Db3", "Ab5"}},
    {67, "Tenor Sax",                       {"Ab2", "Eb5"}},
    {68, "Baritone Sax",                    {"Db2", "Ab4"}},
    {69, "Oboe",                            {"Bb3", "G6"}},
    {70, "English Horn",                    {"E3",  "A5"}},
    {71, "Bassoon",                         {"Bb1", "E5"}},
    {72, "Clarinet",                        {"D3",  "Bb6"}},
    // Pipe
    {73, "Piccolo",                         {"D5",  "C8"}},
    {74, "Flute",                           {"C4",  "C7"}},
    {75, "Recorder",                        {"C-1", "G9"}},
    {76, "Pan Flute",                       {"C-1", "G9"}},
    {77, "Blown bottle",                    {"C-1", "G9"}},
    {78, "Shakuhachi",                      {"C-1", "G9"}},
    {79, "Whistle",                         {"C-1", "G9"}},
    {80, "Ocarina",                         {"C-1", "G9"}},
    // Synth Lead
    {81, "Lead 1 (square)",                 {"C0",  "C8"}},
    {82, "Lead 2 (sawtooth)",               {"C0",  "C8"}},
    {83, "Lead 3 (calliope)",               {"C0",  "C8"}},
    {84, "Lead 4 (chiff)",                  {"C0",  "C8"}},
    {85, "Lead 5 (charang)",                {"C0",  "C8"}},
    {86, "Lead 6 (voice)",                  {"C0",  "C8"}},
    {87, "Lead 7 (fifths)",                 {"C0",  "C8"}},
    {88, "Lead 8 (bass + lead)",            {"C0",  "C8"}},
    // Synth Pad
    {89, "Pad 1 (new age)",                 {"C0",  "C8"}},
    {90, "Pad 2 (warm)",                    {"C0",  "C8"}},
    {91, "Pad 3 (polysynth)",               {"C0",  "C8"}},
    {92, "Pad 4 (choir)",                   {"C2",  "G5"}},
    {93, "Pad 5 (bowed)",                   {"C0",  "C8"}},
    {94, "Pad 6 (metallic)",                {"C0",  "C8"}},
    {95, "Pad 7 (halo)",                    {"C0",  "C8"}},
    {96, "Pad 8 (sweep)",                   {"C0",  "C8"}},
    // Synth Effects
    {97, "FX 1 (rain)",                     {"C0",  "C8"}},
    {98, "FX 2 (soundtrack)",               {"C0",  "C8"}},
    {99, "FX 3 (crystal)",                  {"C0",  "C8"}},
    {100, "FX 4 (atmosphere)",              {"C0",  "C8"}},
    {101, "FX 5 (brightness)",              {"C0",  "C8"}},
    {102, "FX 6 (goblins)",                 {"C0",  "C8"}},
    {103, "FX 7 (echoes)",                  {"C0",  "C8"}},
    {104, "FX 8 (sci-fi)",                  {"C0",  "C8"}},
    // Ethnic
    {105, "Sitar",                          {"C2",  "C6"}},
    {106, "Banjo",                          {"D3",  "C6"}},
    {107, "Shamisen",                       {"E3",  "C6"}},
    {108, "Koto",                           {"C2",  "C6"}},
    {109, "Kalimba",                        {"G3",  "G6"}},
    {110, "Bag pipe",                       {"C3",  "C6"}},
    {111, "Fiddle",                         {"G3",  "C7"}},
    {112, "Shanai",                         {"Bb3", "G6"}},
    // Percussive
    {113, "Tinkle Bell",                    {"C-1", "G9"}},
    {114, "Agogô",                          {"C-1", "G9"}},
    {115, "Steel Drums",                    {"C-1", "G9"}},
    {116, "Woodblock",                      {"C-1", "G9"}},
    {117, "Taiko Drum",                     {"C-1", "G9"}},
    {118, "Melodic Tom",                    {"C-1", "G9"}},
    {119, "Synth Drum",                     {"C-1", "G9"}},
    {120, "Reverse Cymbal",                 {"C-1", "G9"}},
    // Sound Effects
    {121, "Guitar Fret Noise",              {"C-1", "G9"}},
    {122, "Breath Noise",                   {"C-1", "G9"}},
    {123, "Seashore",                       {"C-1", "G9"}},
    {124, "Bird Tweet",                     {"C-1", "G9"}},
    {125, "Telephone Ring",                 {"C-1", "G9"}},
    {126, "Helicopter",                     {"C-1", "G9"}},
    {127, "Applause",                       {"C-1", "G9"}},
    {128, "Gunshot",                        {"C-1", "G9"}},
    // Idiophones
    {129, "Channel 10 percussion",          {"Eb1", "Eb6"}}
};

File: GeneralMidi.py
#!/usr/bin/env python3
"""TextMIDITools: TextMidiFormEdit.py general_midi data with instrument names and codes."""
# TextMIDITools Version 1.1.0
# TextMidiFormEdit.py 1.0
# Copyright © 2025 Thomas E. Janzen
# License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
# This is free software: you are free to change and redistribute it.
# There is NO WARRANTY, to the extent permitted by law.
#
# Some instrument ranges were found on https://en.wikipedia.org or guessed.
# Corrections on unusual instruments are welcome.

GENERAL_MIDI_INSTRUMENT_DICT = {
     'Acoustic_Grand_Piano'            : (  1, 'A0',  'C8'),
     'Bright_Acoustic_Piano'           : (  2, 'A0',  'C8'),
     'Electric_Grand_Piano'            : (  3, 'E1',  'E7'),
     'Honky-tonk_Piano'                : (  4, 'A0',  'C8'),
     'Electric_Piano_1(Fender_Rhodes)' : (  5, 'A0',  'C8'),
     'Electric_Piano_2(DX-7_EP)'       : (  6, 'A0',  'C8'),
     'Harpsichord'                     : (  7, 'F1',  'F6'),
     'Clavi'                           : (  8, 'F1',  'F6'),
     # Chromatic Percussion
     'Celesta'                         : (  9, 'C4',  'C8'),
    'Glockenspiel'                     : ( 10, 'F5',  'C8'),
    'Music_Box'                        : ( 11, 'F6',  'C8'),
    'Vibraphone'                       : ( 12, 'F3',  'F6'),
    'Marimba'                          : ( 13, 'C2',  'C7'),
    'Xylophone'                        : ( 14, 'C3',  'C7'),
    'Tubular Bells'                    : ( 15, 'C4',  'G5'),
    'Dulcimer'                         : ( 16, 'C3',  'C5'),
    # Organ
    'Drawbar_Organ'                    : ( 17, 'A0',  'C8'),
    'Percussive_Organ'                 : ( 18, 'A0',  'C8'),
    'Rock_Organ'                       : ( 19, 'A0',  'C8'),
    'Church_Organ'                     : ( 20, 'A0',  'C8'),
    'Reed_Organ'                       : ( 21, 'C2',  'C6'),
    'Accordion'                        : ( 22, 'E1',  'D6'),
    'Harmonica'                        : ( 23, 'C3',  'G5'),
    'Tango Accordion'                  : ( 24, 'E1',  'D6'),
    # Guitar
    'Acoustic_Guitar_(nylon)'          : ( 25, 'E2',  'D5'),
    'Acoustic_Guitar_(steel)'          : ( 26, 'E2',  'D5'),
    'Electric_Guitar_(jazz)'           : ( 27, 'E2',  'D5'),
    'Electric_Guitar_(clean)'          : ( 28, 'E2',  'D5'),
    'Electric_Guitar_(muted)'          : ( 29, 'E2',  'D5'),
    'Overdriven_Guitar'                : ( 30, 'E2',  'D5'),
    'Distortion_Guitar'                : ( 31, 'E2',  'D5'),
    'Guitar_Harmonics'                 : ( 32, 'E3',  'D6'),
    # Bass
    'Acoustic_Bass'                    : ( 33, 'E1',  'E4'),
    'Electric_Bass_(finger)'           : ( 34, 'E1',  'E4'),
    'Electric_Bass_(pick)'             : ( 35, 'E1',  'E4'),
    'Fretless_Bass'                    : ( 36, 'E1',  'E4'),
    'Slap_Bass_1'                      : ( 37, 'E1',  'E4'),
    'Slap_Bass_2'                      : ( 38, 'E1',  'E4'),
    'Synth_Bass_1'                     : ( 39, 'E1',  'E4'),
    'Synth_Bass_2'                     : ( 40, 'E1',  'E4'),
    # Strings
    'Violin'                           : ( 41, 'G3',  'C7'),
    'Viola'                            : ( 42, 'C3',  'D7'),
    'Cello'                            : ( 43, 'C2',  'A5'),
    'Contrabass'                       : ( 44, 'E1',  'G4'),
    'Tremolo_Strings'                  : ( 45, 'E1',  'C7'),
    'Pizzicato_Strings'                : ( 46, 'E1',  'C7'),
    'Orchestral_Harp'                  : ( 47, 'B0',  'G#7'),
    'Timpani'                          : ( 48, 'C2',  'D4'),
    # Ensemble
    'String_Ensemble_1'                : ( 49, 'C0',  'C8'),
    'String_Ensemble_2'                : ( 50, 'C0',  'C8'),
    'Synth_Strings_1'                  : ( 51, 'C0',  'C8'),
    'Synth_Strings_2'                  : ( 52, 'C0',  'C8'),
    'Choir_Aahs'                       : ( 53, 'D2',  'C6'),
    'Voice_Oohs'                       : ( 54, 'D2',  'C6'),
    'Synth_Voice'                      : ( 55, 'D2',  'C6'),
    'Orchestra_Hit'                    : ( 56, 'C0',  'C8'),
    # Brass
    'Trumpet'                          : ( 57, 'F#3', 'C6'),
    'Trombone'                         : ( 58, 'C2',  'C5'),
    'Tuba'                             : ( 59, 'Eb2', 'A4'),
    'Muted_Trumpet'                    : ( 60, 'F#3', 'C6'),
    'French_Horn'                      : ( 61, 'C2',  'C6'),
    'Brass_Section'                    : ( 62, 'C2',  'C6'),
    'Synth_Brass_1'                    : ( 63, 'C2',  'C6'),
    'Synth_Brass_2'                    : ( 64, 'C2',  'C6'),
    # Reed
    'Soprano_Sax'                      : ( 65, 'Ab3', 'Eb6'),
    'Alto_Sax'                         : ( 66, 'Db3', 'Ab5'),
    'Tenor_Sax'                        : ( 67, 'Ab2', 'Eb5'),
    'Baritone_Sax'                     : ( 68, 'Db2', 'Ab4'),
    'Oboe'                             : ( 69, 'Bb3', 'G6'),
    'English_Horn'                     : ( 70, 'E3',  'A5'),
    'Bassoon'                          : ( 71, 'Bb1', 'E5'),
    'Clarinet'                         : ( 72, 'D3',  'Bb6'),
    # Pipe
    'Piccolo'                          : ( 73, 'D5',  'C8'),
    'Flute'                            : ( 74, 'C4',  'C7'),
    'Recorder'                         : ( 75, 'C-1', 'G9'),
    'Pan_Flute'                        : ( 76, 'C-1', 'G9'),
    'Blown_bottle'                     : ( 77, 'C-1', 'G9'),
    'Shakuhachi'                       : ( 78, 'C-1', 'G9'),
    'Whistle'                          : ( 79, 'C-1', 'G9'),
    'Ocarina'                          : ( 80, 'C-1', 'G9'),
    # Synth Lead
    'Lead_1_(square)'                  : ( 81, 'C0',  'C8'),
    'Lead_2_(sawtooth)'                : ( 82, 'C0',  'C8'),
    'Lead_3_(calliope)'                : ( 83, 'C0',  'C8'),
    'Lead_4_(chiff)'                   : ( 84, 'C0',  'C8'),
    'Lead_5_(charang)'                 : ( 85, 'C0',  'C8'),
    'Lead_6_(voice)'                   : ( 86, 'C0',  'C8'),
    'Lead_7_(fifths)'                  : ( 87, 'C0',  'C8'),
    'Lead_8_(bass_+_lead)'             : ( 88, 'C0',  'C8'),
    # Synth Pad
    'Pad_1_(new_age)'                  : ( 89, 'C0',  'C8'),
    'Pad_2_(warm)'                     : ( 90, 'C0',  'C8'),
    'Pad_3_(polysynth)'                : ( 91, 'C0',  'C8'),
    'Pad_4_(choir)'                    : ( 92, 'C2',  'G5'),
    'Pad_5_(bowed)'                    : ( 93, 'C0',  'C8'),
    'Pad_6_(metallic)'                 : ( 94, 'C0',  'C8'),
    'Pad_7_(halo)'                     : ( 95, 'C0',  'C8'),
    'Pad_8_(sweep)'                    : ( 96, 'C0',  'C8'),
    # Synth Effects
    'FX_1_(rain)'                      : ( 97, 'C0',  'C8'),
    'FX_2_(soundtrack)'                : ( 98, 'C0',  'C8'),
    'FX_3_(crystal)'                   : ( 99, 'C0',  'C8'),
    'FX_4_(atmosphere)'                : (100, 'C0',  'C8'),
    'FX_5_(brightness)'                : (101, 'C0',  'C8'),
    'FX_6_(goblins)'                   : (102, 'C0',  'C8'),
    'FX_7_(echoes)'                    : (103, 'C0',  'C8'),
    'FX_8_(sci-fi)'                    : (104, 'C0',  'C8'),
    # Ethnic
    'Sitar'                            : (105, 'C2',  'C6'),
    'Banjo'                            : (106, 'D3',  'C6'),
    'Shamisen'                         : (107, 'E3',  'C6'),
    'Koto'                             : (108, 'C2',  'C6'),
    'Kalimba'                          : (109, 'G3',  'G6'),
    'Bag pipe'                         : (110, 'C3',  'C6'),
    'Fiddle'                           : (111, 'G3',  'C7'),
    'Shanai'                           : (112, 'Bb3', 'G6'),
    # Percussive
    'Tinkle Bell'                      : (113, 'C-1', 'G9'),
    'Agogô'                            : (114, 'C-1', 'G9'),
    'Steel_Drums'                      : (115, 'C-1', 'G9'),
    'Woodblock'                        : (116, 'C-1', 'G9'),
    'Taiko_Drum'                       : (117, 'C-1', 'G9'),
    'Melodic_Tom'                      : (118, 'C-1', 'G9'),
    'Synth_Drum'                       : (119, 'C-1', 'G9'),
    'Reverse_Cymbal'                   : (120, 'C-1', 'G9'),
    # Sound Effects
    'Guitar_Fret_Noise'                : (121, 'C-1', 'G9'),
    'Breath_Noise'                     : (122, 'C-1', 'G9'),
    'Seashore'                         : (123, 'C-1', 'G9'),
    'Bird_Tweet'                       : (124, 'C-1', 'G9'),
    'Telephone_Ring'                   : (125, 'C-1', 'G9'),
    'Helicopter'                       : (126, 'C-1', 'G9'),
    'Applause'                         : (127, 'C-1', 'G9'),
    'Gunshot'                          : (128, 'C-1', 'G9'),
}
GM_PROGRAM_INDEX = 0
GM_LOW_PITCH_INDEX = 1
GM_HIGH_PITCH_INDEX = 2
